### PR TITLE
MatrixSymbol first arg is a Symbol (and other matrix expressions cleanup)

### PR DESCRIPTION
**Pull request by asmeurer**
The main change here is to make it so that `MatrixSymbol.args[0]` is a Symbol instead of a string. `MatrixSymbol.name` is still a string. (you can still create it with a string, it just gets converted to a Symbol in the .args automatically)

#### References to other Issues or PRs

I still need to do an issue tracker sweep to see what kinds of issues are affected. I know at least the printing issue from https://github.com/sympy/sympy/issues/15120 is fixed here. 

#### Brief description of what is fixed or changed

See the commit messages for the minor cleanups.

#### Other comments

This has been a huge issue ever since the creation of MatrixSymbol. The problem is that non-Basic args break any code that assumes args are Basic. Most likely there is a lot of code that special-cases MatrixSymbol that can be removed here. I still need to see if any tests fail first. 

This is a backwards incompatible break. I don't think there's any way to do a deprecation.  I've left the type of `MatrixSymbol.name` intact to keep as much compatibility as possible. 

## TODO: 

- [ ] Add tests for
  - [ ] #15120
  - [ ] #6249
  - [x] https://github.com/sympy/sympy/issues/6913 (works in master but make sure it is tested)
  - [x] https://github.com/sympy/sympy/issues/2750 (works in master)
  - [ ] https://github.com/sympy/sympy/issues/7842 (works in master)
  - [ ] https://github.com/sympy/sympy/issues/9817 (str related issue works in master)
  - [ ] https://github.com/sympy/sympy/issues/11948 (not sure the exact issue but it seems to work in master)
  - [ ] https://github.com/sympy/sympy/pull/13453 (the recursion error is gone, but there's a fixed printing error from master)
  - [ ] Check if https://github.com/sympy/sympy/pull/14424 is still necessary (c.f. https://github.com/sympy/sympy/issues/14421, which is broken in master and fixed here)
  - [ ] https://github.com/sympy/sympy/issues/15050 (no longer errors but doesn't give a solution)
  - [ ] Check for other issues
- [x] See if removing the `isinstance(..., string_types)` fails any tests. https://github.com/sympy/sympy/pull/15121#issuecomment-423315973
  - [x] Check for other special casing of string args
- [x] Fix `exp(A)*exp(B)` https://github.com/sympy/sympy/pull/15121#issuecomment-442182341
- [x] XFAIL incorrect matrix derivative test https://github.com/sympy/sympy/pull/15121#discussion_r242716398
#### Release Notes

* core
  * `count_ops` no longer handles objects with strings in their `.args` (all elements of `.args` for SymPy objects should be `Basic` subclasses).
* matrices
  * `MatrixSymbol.args[0]` is now a `Symbol` instead of a string. `MatrixSymbol.name` is still a string.
* printing
  * srepr now correctly prints subclasses of `Add` and `Mul`.

### CODE REVIEW

File: sympy/matrices/expressions/tests/test_derivatives.py [hunk 1]
```diff
@@ -86,3 +86,8 @@ def test_matrix_derivative_with_inverse():
     # Cookbook example 64:
     expr = Trace(Inverse(X + A))
     assert expr.diff(X) == -(Inverse(X + A)*Inverse(X + A)).T
+
+def test_trace():
+    # This should probably simplify more
+    expr = Trace(A)*A
+    assert expr.diff(A) == Derivative(expr, A)
```
[the remark this answers is not in the document]
Comment (Upabjojr) [reply sympy/matrices/expressions/tests/test_derivatives.py]: What do you mean?

[the remark this answers is not in the document]
Comment (asmeurer) [reply sympy/matrices/expressions/tests/test_derivatives.py]: Actually never mind. It is used, but only internally. But the other question remains, should `diff(Trace(A)*A, A)` simplify? I suppose the answer is probably no, given that diff(A, A) doesn't simplify. 

Also, unrelated, but `diff(Trace(A), A)` gives `1`, but the cookbook says it should be `I`.

File: sympy/matrices/expressions/matexpr.py [hunk 2]
```diff
@@ -596,12 +596,14 @@ class MatrixElement(Expr):
...
-        n, m = map(sympify, (n, m))
+        n, m = map(_sympify, (n, m))
         from sympy import MatrixBase
         if isinstance(name, (MatrixBase,)):
             if n.is_Integer and m.is_Integer:
                 return name[n, m]
-        name = sympify(name)
+        if isinstance(name, string_types):
+            name = Symbol(name)
+        name = _sympify(name)
```
Comment (smichr) [inline sympy/matrices/expressions/matexpr.py]: Should that `name = _sy...` be
```python
elif not isinstance(name, Symbol):
    raise TypeError('name must be string of Symbol')
```

File: sympy/matrices/expressions/matexpr.py [hunk 3]
```diff
@@ -658,7 +660,9 @@ class MatrixSymbol(MatrixExpr):
     _diff_wrt = True
 
     def __new__(cls, name, n, m):
-        n, m = sympify(n), sympify(m)
+        n, m = _sympify(n), _sympify(m)
+        if isinstance(name, string_types):
+            name = Symbol(name)
```
Comment (smichr) [inline sympy/matrices/expressions/matexpr.py]: elif could be added here, too?

File: sympy/matrices/expressions/matexpr.py [hunk 4]
```diff
@@ -596,12 +596,14 @@ class MatrixElement(Expr):
...
 
     def __new__(cls, name, n, m):
-        n, m = map(sympify, (n, m))
+        n, m = map(_sympify, (n, m))
         from sympy import MatrixBase
         if isinstance(name, (MatrixBase,)):
             if n.is_Integer and m.is_Integer:
                 return name[n, m]
-        name = sympify(name)
+        if isinstance(name, string_types):
```
Comment (smichr) [inline sympy/matrices/expressions/matexpr.py]: and this could be `elif` since it's not a MatrixBase.

**Comment by asmeurer** [comment]
All the simplify functions (https://github.com/sympy/sympy/issues/6249) now work.

**Comment by asmeurer** [comment]
There are some lines in the core like

https://github.com/sympy/sympy/blob/3d0e2ec4dcb7653f7fac039cc585ae81dc7251c2/sympy/core/function.py#L2632-L2634

https://github.com/sympy/sympy/blob/3d0e2ec4dcb7653f7fac039cc585ae81dc7251c2/sympy/core/function.py#L2723-L2725

which were definitely put there for MatrixSymbol. But there may be other things as well. I can try removing them and seeing what tests fail.

**Comment by asmeurer** [comment]
When searching for issues for this, I also found many issues like https://github.com/sympy/sympy/issues/11521 (`AttributeError: 'MatrixSymbol' object has no attribute '_diff_wrt'`) that are now fixed. Someone should go through and make sure they are tested and close them.

**Comment by asmeurer** [comment]
My TODO list above is I believe extensive. What remains to be done here is to go through each issue and make sure it is tested here, and close it once this is merged.

**Comment by asmeurer** [comment]
Tests pass without those checks in count_ops. Should we remove them?

**Comment by asmeurer** [comment]
It seems there is a bug here:

```pytb
>>> A = MatrixSymbol('A', 2, 2)
>>> B = MatrixSymbol('B', 2, 2)
>>> exp(A)*exp(B)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "./sympy/core/decorators.py", line 91, in __sympifyit_wrapper
    return func(a, b)
  File "./sympy/core/decorators.py", line 132, in binary_op_wrapper
    return func(self, other)
  File "./sympy/core/expr.py", line 140, in __mul__
    return Mul(self, other)
  File "./sympy/core/operations.py", line 45, in __new__
    c_part, nc_part, order_symbols = cls.flatten(args)
  File "./sympy/core/mul.py", line 354, in flatten
    o12 = b1 ** new_exp
  File "./sympy/core/expr.py", line 154, in __pow__
    return self._pow(other)
  File "./sympy/core/decorators.py", line 91, in __sympifyit_wrapper
    return func(a, b)
  File "./sympy/core/decorators.py", line 131, in binary_op_wrapper
    return f(self)
  File "./sympy/matrices/expressions/matexpr.py", line 24, in __sympifyit_wrapper
    return func(a, b)
  File "./sympy/core/decorators.py", line 132, in binary_op_wrapper
    return func(self, other)
  File "./sympy/matrices/expressions/matexpr.py", line 145, in __rpow__
    raise NotImplementedError("Matrix Power not defined")
NotImplementedError: Matrix Power not defined
```

**Comment by jksuom** [comment]
File "./sympy/core/mul.py", line 354, in flatten
       o12 = b1 ** new_exp

This tries to compute `E**(A + B)` which is wrong. On the previous line, it is checked that `not new_exp.is_Add`, but it does not help as `A + B` is of type `MatAdd`.

**Comment by jksuom** [comment]
`exp(A)*exp(B)` should probably not be handled by core. Maybe `exp(A)` should be a matrix.

**Comment by asmeurer** [comment]
This is related to https://github.com/sympy/sympy/issues/15551

**Comment by asmeurer** [comment]
It seems the merge with master fixed the problem.

File: sympy/matrices/expressions/tests/test_derivatives.py [hunk 5]
```diff
@@ -101,6 +101,9 @@ def test_matrix_derivative_vectors_and_scalars():
 
 def test_matrix_derivatives_of_traces():
 
+    expr = Trace(A)*A
+    assert expr.diff(A) == Trace(A)*Identity(k) + A.T
```
Comment (asmeurer) [inline sympy/matrices/expressions/tests/test_derivatives.py]: @Upabjojr can you confirm if this is correct?

**Comment by asmeurer** [comment]
Some of the simplification functions that worked before don't work any more, as a result of https://github.com/sympy/sympy/pull/15373. I am digging into the reasons. It looks like some code ends up creating Add or Mul instead of MatAdd or MatMul on matrix expressions.

File: sympy/matrices/expressions/tests/test_derivatives.py (hunk 5 above)
Comment (Upabjojr) [reply sympy/matrices/expressions/tests/test_derivatives.py]:
```
Trace(A)*A) ===> A_{ii} A_{ij}
```
now
```
diff(A_ii A_jk, A_mn) = d_mi d_ni A_jk + A_ii d_mj d_nk = d_mn A_jk + A_ii d_mj d_nk
```
There is a problem with this result, as it should probably be a higher-rank array. MatrixCalculus.org is indeed unable to compute it.

Comment (Upabjojr) [reply sympy/matrices/expressions/tests/test_derivatives.py]: I'd say the result is wrong, it's omitting a higher dimensional identity.

Comment (Upabjojr) [reply sympy/matrices/expressions/tests/test_derivatives.py]: https://github.com/sympy/sympy/issues/15667

**Comment by RituRajSingh878** [comment]
@asmeurer  I would like to work on these tasks.I think here we have to add test.

**Comment by asmeurer** [comment]
After https://github.com/sympy/sympy/pull/15948, almost all simplifying functions will raise NotImplementedError. They will need to be adapted manually to work with matrix expressions.

**Comment by asmeurer** [comment]
Should we try to get this in the 1.4 release? It would be nice to have it, but it also isn't critical.

**Comment by asmeurer** [comment]
Another place that handles non-Basic args https://github.com/sympy/sympy/blob/f499cbbc188d777be2998b01e48a3d8b64f1ea10/sympy/core/exprtools.py#L1093

**Comment by cbm755** [comment]
I ran my downstream Octave project test set on this without any additional failures, so I guess that's a partial +1 for merging into 1.4.  That is, it sounds reasonable, doesn't break anything for me, but I haven't carefully reviewed.

**Comment by asmeurer** [comment]
Any objections to merging this into the 1.4 release branch?

**Comment by smichr** [comment]
> all elements of .args for SymPy objects should be Basic subclasses

This is not enforced by Matrices, but they are not a Basic instance. I was surpriced that `laplace_transform` send back a matrix with lists as entries. In the work I have done recently with matrices I have tried to be as lenient as possible for Matrix contents.

BTW, I'm +1 to committing this. I see you have several TODOs so did not commit this now. I'm not sure if you want to finish those before this goes in or not.

**Comment by asmeurer** [comment]
I'll double check them but a lot of them were things that I was going to add tests for but they now don't work after #15948.

**Pull request merged**